pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java

```diff
@@ -73,20 +84,19 @@ public FilterPlanNode(IndexSegment indexSegment, QueryContext queryContext) {
   @Override
   public BaseFilterOperator run() {
     FilterContext filter = _queryContext.getFilter();
-    ThreadSafeMutableRoaringBitmap validDocIds = _indexSegment.getValidDocIds();
-    boolean applyValidDocIds = validDocIds != null && !QueryOptionsUtils.isSkipUpsert(_queryContext.getQueryOptions());
+    boolean applyValidDocIds = _validDocIds != null && !QueryOptionsUtils.isSkipUpsert(_queryContext.getQueryOptions());
     if (filter != null) {
       BaseFilterOperator filterOperator = constructPhysicalOperator(filter);
       if (applyValidDocIds) {
         BaseFilterOperator validDocFilter =
-            new BitmapBasedFilterOperator(validDocIds.getMutableRoaringBitmap(), false, _numDocs);
+                new BitmapBasedFilterOperator(_validDocIds, false, _numDocs);
```

Comment:
Are you using the pinot code style?
https://docs.pinot.apache.org/developers/developers-and-contributors/code-setup#setup-ide

Reply:
Ah my bad let me fix that.

Reply:
With regard to first question: what I am trying to fix is known ids disappearing temporarily from the results. Say there are 1000 positions with ids 1-1000 and they are all published at the start of the day and then updated during the day. Whenever I query regardless of what upsets are in place I would like to get exactly 1000 items back - one for each id.

What is currently happening is that numDocs is snapshotted - say 50,000 during the middle of the day - then concurrently some id, is updated and the new update is in row 50,001 and the bitmap is updated - lastly the validDocId bitmap is snapshotted. That means when the query runs row 50,001 is excluded due to the numDoc snapshot and the result is only 999 docs - one missing.

This kind of case - missing one or two records that were previously available - happens fairly reliable and is the result of snapshotting the total docs first while concurrent operations can still happen on the validity bitmap.